### PR TITLE
Low latency glm5

### PR DESCRIPTION
**Pull request by yijingl-nvidia**
## 📌 Description

Codex generated summary below as I don't have enough bandwidth to manually check its outcome. But I can at least guarantee the added kernels are real! Take the rest of the content with a grain of salt.

This PR adds a specialized low-latency fused MoE path for GLM-5 inference on NVIDIA Blackwell GPUs.

The implementation targets small-token decode workloads and uses a simplified two-stage TMA design:

1. Fused routing, FP8 activation quantization, expert up-projection, and SwiGLU.
2. Expert down-projection with routed/shared expert reduction.

The PR includes:

- `glm5_low_latency_moe` execution API.
- `prepare_glm5_low_latency_moe_weights` weight preparation API.
- `alloc_glm5_low_latency_moe_workspace` workspace allocation API.
- Blackwell SM100 kernel implementations.
- AOT and JIT integration.
- FlashInfer trace-template integration.
- Correctness tests using saved GLM-5 TP8 tensors.
- An 8-GPU benchmark comparing the new path with the runnable general-purpose CUTLASS BF16 MoE path.

The operator covers the local MoE computation. The router GEMM and TP all-reduce remain outside the operator.

### Performance

Measured on 8 NVIDIA B200 GPUs using TP8 checkpoint shards, 4 input tokens, 20 warm-up iterations, and 100 measured iterations:

| Path | Mean latency | Median latency |
|---|---:|---:|
| CUTLASS BF16, kernel only | 131.988 µs | 130.218 µs |
| CUTLASS BF16 with external routing | 230.653 µs | 228.078 µs |
| GLM-5 low-latency fused MoE | 41.736 µs | 41.202 µs |

The new path is:

- 3.162× faster than the CUTLASS kernel-only measurement.
- 5.526× faster than CUTLASS including its external routing operations.

Checkpoint block-FP8 weights are dequantized to BF16 once outside the CUTLASS timing because the existing CUTLASS block-FP8 MoE path is not implemented for SM100.

### Correctness

All eight ranks passed comparison with the saved PyTorch reference using a maximum absolute-error threshold of `1e-3`.

- Maximum GLM-5 error across ranks: `4.08e-4`
- Maximum CUTLASS error across ranks: `5.22e-4`
- Maximum difference between the two paths: `1.14e-4`

## 🔍 Related Issues

N/A

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Added a shape-specialized GLM5 low-latency FP8 MoE backend for supported GPUs.
  * Integrated GLM5 execution into the unified MoE API with automatic configuration, routing, weight preparation, and reusable buffers.
  * Added support for fused shared experts and MiniMax2 routing.
  * Added JIT/AOT build support for the new backend.
  * Added benchmarking for correctness, latency, profiling, and multi-rank performance.

* **Documentation**
  * Updated GLM5 integration guidance with unified API examples and configuration details.

* **Tests**
  * Added unified API validation and replay coverage for GLM5 workloads.

### CODE REVIEW

File: benchmarks/bench_glm5_low_latency_moe.py [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+        dist.all_gather(gathered, local_stats)
+        if rank == 0:
+            all_stats = torch.stack(gathered).cpu().reshape(world_size, len(names), 3)
+            means = all_stats[:, :, 0].mean(dim=0)
+            for index, name in enumerate(names):
+                print(
+                    f"all_ranks={world_size} tokens={args.tokens} path={name} "
+                    f"mean_us={means[index].item():.3f} "
+                    f"median_us={all_stats[:, index, 1].mean().item():.3f} "
+                    f"rank_mean_range_us=[{all_stats[:, index, 0].min().item():.3f}, "
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply benchmarks/bench_glm5_low_latency_moe.py:382]: Resolved by agent

File: csrc/fused_moe/glm5/glm5_fused_expert_down.cu [hunk 2]
```diff
@@ -0,0 +1,1592 @@
...
+    cuuint32_t elem_stride[3] = {1u, 1u, 1u};
+
+    *out_err = cuTensorMapEncodeTiled(&map, CU_TENSOR_MAP_DATA_TYPE_UINT8,
+        /*rank=*/3, base_ptr, global_dim, global_stride, box_dim, elem_stride, CU_TENSOR_MAP_INTERLEAVE_NONE,
+        CU_TENSOR_MAP_SWIZZLE_128B, CU_TENSOR_MAP_L2_PROMOTION_L2_256B, CU_TENSOR_MAP_FLOAT_OOB_FILL_NONE);
+    return map;
+}
+
+constexpr size_t kWDownTmaDescCacheCap = 256;
+constexpr int kMaxCudaDevicesForSmemAttr = 16;
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply csrc/fused_moe/glm5/glm5_fused_expert_down.cu]: Resolved by agent

File: docs/design_docs/glm5_low_latency_moe.md [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+layer. It fuses no-aux sigmoid top-k routing, activation quantization, the
+gate/up projections, SwiGLU, the down projections, and routed/shared reduction
+into two CUDA launches. It is intended for tensor-parallel decode with one to
+four tokens per rank.
+
+The CUDA implementation is adapted from the `fuse_more_kernels` branch of
+[`yijingl-nvidia/TensorRT-LLM`](https://github.com/yijingl-nvidia/TensorRT-LLM/tree/fuse_more_kernels),
+at source revision `73dd714d5`. The port replaces the Torch custom-op boundary
+with TVM FFI, uses caller-owned outputs and reusable workspaces, and integrates
+with FlashInfer's JIT, AOT, trace, test, and benchmark infrastructure.
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply docs/design_docs/glm5_low_latency_moe.md]: Fixed

File: tests/moe/test_glm5_low_latency_moe.py [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+    dump_env = os.environ.get("FLASHINFER_GLM5_LOW_LATENCY_MOE_DUMP_DIR")
+    if not dump_env:
+        pytest.skip(
+            "set FLASHINFER_GLM5_LOW_LATENCY_MOE_DUMP_DIR to run the GLM5 dump replay"
+        )
+
+    rank = int(os.environ["RANK"])
+    world_size = int(os.environ["WORLD_SIZE"])
+    local_rank = int(os.environ["LOCAL_RANK"])
+    assert world_size == 8
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply tests/moe/test_glm5_low_latency_moe.py:210]: fixed by agent

File: tests/trace/example.py [hunk 5]
```diff
@@ -34,6 +34,7 @@
 gemma_rmsnorm_h4608.json
 gelu_and_mul_h16384.json
 gelu_tanh_and_mul_h16384.json
+glm5_low_latency_moe_topk8_h6144_i256.json
```
[the remark this answers is not in the document]
Comment (yijingl-nvidia) [reply tests/trace/example.py]: The file is already committed
